x-pack/plugins/apm/public/hooks/table_sort_pagination/use_table_sort_pagination.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CriteriaWithPagination,
+  EuiTableSortingType,
+  Pagination,
+} from '@elastic/eui';
+import { orderBy } from 'lodash';
+import { useMemo } from 'react';
+import uuid from 'uuid';
+import { useApmParams } from '../use_apm_params';
+
+export interface TableSortPaginationProps<T extends any[]> {
+  items: T;
+  // totalItemCount is not necessary here
+  initialPagination: Omit<Pagination, 'totalItemCount'>;
+  initialSort?: EuiTableSortingType<T[0]>;
+}
+
+type Props<T extends any[]> = TableSortPaginationProps<T> & {
+  tableOptions: CriteriaWithPagination<T[0]>;
+  onTableChange: (criteriaWithPagination: CriteriaWithPagination<T[0]>) => void;
+};
+
+const PAGE_SIZE_OPTIONS = [10, 25, 50];
+
+export function useTableSortAndPagination<T extends any[]>({
+  items,
+  initialPagination,
+  initialSort = {},
+  tableOptions,
+  onTableChange,
+}: Props<T>): {
+  onTableChange: (criteriaWithPagination: CriteriaWithPagination<T[0]>) => void;
+  tableSort?: EuiTableSortingType<T[0]>;
+  tablePagination: Pagination;
+  tableItems: any[];
+  totalItems: number;
+  requestId: string;
+} {
+  const { query } = useApmParams('/*');
```

Comment:
this is an anti-pattern and should be avoided. It causes us to lose type-safety, in the sense that there is no longer a guarantee that the consuming component has access to the necessary query parameters. As an example, a route might not have defined `offset` in its route params, silently breaking the expected behaviour.

Reply:
How can I overcome this then?

>As an example, a route might not have defined offset in its route params, silently breaking the expected behaviour.

How would that break? At most, it will be `undefined` right?

Reply:
In that case, it won't actually use `offset`, and we won't get a type error for it. Usually the approach I take is to pass query parameters explicitly to a hook or component from the consumer. So you'd have something like (off the top of my head):

```ts

const { query } = useApmParams('/services/{serviceName}/errors');

const { offset, comparisonEnabled: urlComparisonEnabled } = query;

const comparisonEnabled = getComparisonEnabled({ urlComparisonEnabled });

const pagination = useTablePagination({
	offset,
	comparisonEnabled,
});
```

Reply:
I don't think it makes much sense IMHO. This looks more like a poor design of the `useApmParams` and to work this out I have to manually pass down all properties to my hook. 

I'd rather specify the list of paths that contains pagination in the URL than doing like that.

Reply:
The other alternative is to explicitly pass the route path to the hook, and ensure the `TypeOf` that route has the parameters we expect.

Reply:
This doesn't look as bad as I thought https://github.com/elastic/kibana/pull/135665/commits/046f2dbda405f64f1adaac821f0c2894b0be00ee#diff-cb9905f21c3369c42d3f24adcd87da14e0b74b0968c57e724d50a2eabcf672baR110 😅